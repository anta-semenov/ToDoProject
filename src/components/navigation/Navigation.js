import React from 'react'
import ImmutablePropTypes from 'react-immutable-proptypes'
import './Navigation.less'
import NavigationGroup from '../navigationGroup/NavigationGroup'
import DropScrollTarget from '../elements/dropScrollTarget/DropScrollTarget'

export default class Navigation extends React.Component {
  constructor(props) {
    super(props)
  }

  render() {
    return (
      <div className='navigation'>
        <DropScrollTarget className='nav-scroll' scrollCallback = {() => {
          if (this._navScrollView.scrollTop > 0) {
            this._navScrollView.scrollTop -=10
          }
        }} />
        <ul className='nav' ref={(ref) => this._navScrollView = ref} >
          {this.props.groups.map((group, index) =>
            <NavigationGroup
              key={index} //We should pass unique identificator for array items
              items={group.items}
              title={group.title}
              onItemClick={this.props.onItemClick}
              type={group.type}
              addNewTitle={group.addNewTitle}
              addNew={this.props.addNew}
              onStopEditing={this.props.onStopEditing}
              changePosition={this.props.changePosition}
            />
          )}
        </ul>
        <DropScrollTarget className='nav-scroll' scrollCallback = {() => {
          if (this._navScrollView.scrollTop < this._navScrollView.scrollHeight - this._navScrollView.clientHeight) {
            this._navScrollView.scrollTop +=10
          }
        }} />
      </div>
    )
  }
}

Navigation.propTypes = {
  groups: React.PropTypes.arrayOf(
    React.PropTypes.shape({
      type: React.PropTypes.string.isRequired,
      items: ImmutablePropTypes.listOf(
        ImmutablePropTypes.contains({
          type: React.PropTypes.string.isRequired,
          title: React.PropTypes.string.isRequired,
          active: React.PropTypes.bool.isRequired,
          id: React.PropTypes.string,
          count: React.PropTypes.number,
          editing: React.PropTypes.bool
        })
      ).isRequired,
      title: React.PropTypes.string,
      addNewTitle: React.PropTypes.string
    })
  ).isRequired,
  onItemClick: React.PropTypes.func.isRequired,
  addNew: React.PropTypes.func,
  onStopEditing: React.PropTypes.func,
  changePosition: React.PropTypes.func
}
