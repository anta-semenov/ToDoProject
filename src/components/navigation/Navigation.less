@import '../../styles/variables.less';

.nav {
  position: fixed;
  top: @size-height-auth-component;
  bottom: 0;
  left: 0;
  width: @size-width-sidebar;
  box-sizing: border-box;
  margin: 0;
  padding: 30px;
  padding-top: 10px;

  list-style: none;
  font-family: 'Fira Sans', sans-serif;
  overflow: auto;
}
