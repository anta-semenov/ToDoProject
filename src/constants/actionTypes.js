/*
 * Project
 */
export const ADD_PROJECT = 'ADD_PROJECT'
export const REMOVE_PROJECT = 'REMOVE_PROJECT'
export const COMPLETE_PROJECT = 'COMPLETE_PROJECT'
export const EDIT_PROJECT = 'EDIT_PROJECT'
export const REPLACE_PROJECT = 'REPLACE_PROJECT'

/*
 * Task
 */
export const ADD_TASK = 'ADD_TASK'
export const REMOVE_TASK = 'REMOVE_TASK'
export const COMPLETE_TASK = 'COMPLETE_TASK'
export const EDIT_TASK = 'EDIT_TASK'
export const REPLACE_TASK = 'REPLACE_TASK'
export const ADD_TASK_TO_PROJECT = 'ADD_TASK_TO_PROJECT'
export const ADD_TASK_CONTEXT = 'ADD_TASK_CONTEXT'
export const REMOVE_TASK_CONTEXT = 'REMOVE_TASK_CONTEXT'
export const SET_TASK_TODAY = 'SET_TASK_TODAY'
export const START_TASK_TRACKING = 'START_TASK_TRACKING'
export const STOP_TASK_TRACKING = 'STOP_TASK_TRACKING'

/*
 * Context
 */
export const ADD_CONTEXT = 'ADD_CONTEXT'
export const REMOVE_CONTEXT = 'REMOVE_CONTEXT'
export const EDIT_CONTEXT = 'EDIT_CONTEXT'
export const SWITCH_TASK_CONTEXT = 'SWITCH_TASK_CONTEXT'
export const REPLACE_CONTEXT = 'REPLACE_CONTEXT'

/*
* UI State
*/
export const SET_SELECTED_SECTION = 'SET_SELECTED_SECTION'
export const SET_SIDEBAR_SIZE = 'SET_SIDEBAR_SIZE'
export const SET_ACTIVE_ITEM = 'SET_ACTIVE_ITEM'
export const SET_EDITING_SECTION = 'SET_EDITING_SECTION'
export const TOGGLE_TASK_COMPLETED_LATENCY = 'TOGGLE_TASK_COMPLETED_LATENCY'
export const CLEAR_COMPLETED_LATENT_TASKS = 'CLEAR_COMPLETED_LATENT_TASKS'
export const TOGGLE_TASK_LATENCY = 'TOGGLE_TASK_LATENCY'
export const CLEAR_LATENT_TASKS = 'CLEAR_LATENT_TASKS'
export const SET_SYNCING = 'SET_SYNCING'
export const SET_OFFLINE = 'SET_OFFLINE'
export const SET_AUTH_STATUS = 'SET_AUTH_STATUS'
export const SET_AUTH_ERROR_MESSAGE = 'SET_AUTH_ERROR_MESSAGE'
export const SET_PROPERTY = 'SET_PROPERTY'

/*
* Common
*/
export const SET_STATE = 'SET_STATE'
export const LOG_OUT = 'LOG_OUT'
export const REQUEST_AUTH = 'REQUEST_AUTH'
export const RECIEVE_AUTH = 'RECIEVE_AUTH'
export const ERROR_AUTH = 'ERROR_AUTH'

export const REQUEST_DATA = 'REQUEST_DATA'
export const RECIEVE_DATA = 'RECIEVE_DATA'
export const ERROR_DATA = 'ERROR_DATA'

export const UNDO = 'UNDO'
export const REDO = 'REDO'
export const CLEAR_UNDO_REDO = 'CLEAR_UNDO_REDO'
