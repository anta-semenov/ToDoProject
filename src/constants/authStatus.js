export const AUTH_IN_PROGRESS = 'AUTH_IN_PROGRESS'
export const AUTH_SUCESS = 'AUTH_SUCESS'
export const AUTH_ERROR = 'AUTH_ERROR'
export const AUTH_NONE = 'AUTH_NONE'
