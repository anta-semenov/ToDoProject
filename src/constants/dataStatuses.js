export const DATA_NONE = 'DATA_NONE'
export const DATA_REQUESTED = 'DATA_REQUESTED'
export const DATA_ERROR = 'DATA_ERROR'
export const DATA_RECIEVED = 'DATA_RECIEVED'
