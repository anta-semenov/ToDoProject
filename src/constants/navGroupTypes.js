export const BASIC = 'BASIC'
export const CONTEXTS = 'CONTEXTS'
export const PROJECTS = 'PROJECTS'
