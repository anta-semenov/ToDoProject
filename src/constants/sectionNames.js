export const INBOX = 'Inbox'
export const NEXT = 'Next'
export const TODAY = 'Today'
export const CONTEXTS = 'Contexts'
export const PROJECTS = 'Projects'
export const SOMEDAY = 'Someday'
