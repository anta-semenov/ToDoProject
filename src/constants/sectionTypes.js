export const PROJECTS = 'PROJECTS'
export const CONTEXTS = 'CONTEXTS'
export const PROJECT = 'PROJECT'
export const CONTEXT = 'CONTEXT'
export const INBOX = 'INBOX'
export const TODAY = 'TODAY'
export const NEXT = 'NEXT'
export const SOMEDAY = 'SOMEDAY'
