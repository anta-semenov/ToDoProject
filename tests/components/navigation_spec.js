import React, { Component } from 'react'
import { mount } from 'enzyme'
import { fromJS } from 'immutable'
import * as sectionTypes from '../../src/constants/sectionTypes'
import * as sectionNames from '../../src/constants/sectionNames'
import * as navGroupTypes from '../../src/constants/navGroupTypes'
import { DATA_RECIEVED } from '../../src/constants/dataStatuses'
import { AUTH_SUCESS } from '../../src/constants/authStatus'

import Navigation from '../../src/components/navigation/Navigation'

describe('Navigation', () => {
  Navigation.__Rewire__(
    'NavigationGroup',
    class extends Component {
      render() {
        return <div className="nav-group" />
      }
    }
  )
  Navigation.__Rewire__(
    'DropScrollTarget',
    class extends Component {
      render() {
        return null
      }
    }
  )
  const testClickOnItem = jest.fn()
  const testAddNew = jest.fn()
  const testStopEditing = jest.fn()

  const testProps = {
    dataStatus: DATA_RECIEVED,
    authStatus: AUTH_SUCESS,
    groups: fromJS([
      {
        type: navGroupTypes.BASIC,
        items: [
          {
            type: sectionTypes.INBOX,
            title: sectionNames.INBOX,
            active: true,
            count: 3
          },
          {
            type: sectionTypes.TODAY,
            title: sectionNames.TODAY,
            active: false,
            count: 1
          },
          {
            type: sectionTypes.NEXT,
            title: sectionNames.NEXT,
            active: false
          }
        ]
      },
      {
        type: navGroupTypes.CONTEXTS,
        title: sectionNames.CONTEXTS,
        items: [
          {
            type: sectionTypes.CONTEXT,
            title: 'Context 1',
            id: 'cf1sobz3s0oc',
            active: true,
            count: 4,
            editing: false
          },
          {
            type: sectionTypes.CONTEXT,
            title: 'Context 2',
            id: 'cf1sobz3s0ol',
            active: false,
            count: 10,
            editing: false
          }
        ],
        onItemClick: testClickOnItem,
        onStopEditing: testStopEditing,
        addNewTitle: '+ context',
        addNew: testAddNew
      }
    ]),
    onItemClick: () => {},
    addNew: () => {},
    onStopEditing: () => {}
  }
  test('Should render correct component tree', () => {
    const navigation = mount(<Navigation {...testProps} />)
    expect(navigation).toMatchSnapshot()
  })
})
