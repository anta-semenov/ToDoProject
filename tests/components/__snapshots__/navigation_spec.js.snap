// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Should render correct component tree 1`] = `
<Navigation
  addNew={[Function]}
  authStatus="AUTH_SUCESS"
  dataStatus="DATA_RECIEVED"
  groups={
    Immutable.List [
      Immutable.Map {
        "type": "BASIC",
        "items": Immutable.List [
          Immutable.Map {
            "type": "inbox",
            "title": "Inbox",
            "active": true,
            "count": 3,
          },
          Immutable.Map {
            "type": "today",
            "title": "Today",
            "active": false,
            "count": 1,
          },
          Immutable.Map {
            "type": "next",
            "title": "Next",
            "active": false,
          },
        ],
      },
      Immutable.Map {
        "type": "CONTEXTS",
        "title": "Contexts",
        "items": Immutable.List [
          Immutable.Map {
            "type": "CONTEXT",
            "title": "Context 1",
            "id": "cf1sobz3s0oc",
            "active": true,
            "count": 4,
            "editing": false,
          },
          Immutable.Map {
            "type": "CONTEXT",
            "title": "Context 2",
            "id": "cf1sobz3s0ol",
            "active": false,
            "count": 10,
            "editing": false,
          },
        ],
        "onItemClick": [MockFunction],
        "onStopEditing": [MockFunction],
        "addNewTitle": "+ context",
        "addNew": [MockFunction],
      },
    ]
  }
  onItemClick={[Function]}
  onStopEditing={[Function]}
>
  <div
    className="navigation"
  >
    <_class2
      className="nav-scroll nav-scroll--up"
      scrollCallback={[Function]}
    />
    <ul
      className="nav"
    >
      <_class
        addNew={[Function]}
        items={
          Immutable.List [
            Immutable.Map {
              "type": "inbox",
              "title": "Inbox",
              "active": true,
              "count": 3,
            },
            Immutable.Map {
              "type": "today",
              "title": "Today",
              "active": false,
              "count": 1,
            },
            Immutable.Map {
              "type": "next",
              "title": "Next",
              "active": false,
            },
          ]
        }
        key="0"
        onItemClick={[Function]}
        onStopEditing={[Function]}
        type="BASIC"
      >
        <div
          className="nav-group"
        />
      </_class>
      <_class
        addNew={[Function]}
        addNewTitle="+ context"
        items={
          Immutable.List [
            Immutable.Map {
              "type": "CONTEXT",
              "title": "Context 1",
              "id": "cf1sobz3s0oc",
              "active": true,
              "count": 4,
              "editing": false,
            },
            Immutable.Map {
              "type": "CONTEXT",
              "title": "Context 2",
              "id": "cf1sobz3s0ol",
              "active": false,
              "count": 10,
              "editing": false,
            },
          ]
        }
        key="1"
        onItemClick={[Function]}
        onStopEditing={[Function]}
        title="Contexts"
        type="CONTEXTS"
      >
        <div
          className="nav-group"
        />
      </_class>
    </ul>
    <_class2
      className="nav-scroll nav-scroll--down"
      scrollCallback={[Function]}
    />
  </div>
</Navigation>
`;
